### PR TITLE
Add unit tests for planner functions

### PR DESCRIPTION
**Pull request by mrivi**
Review only commit 12d8f1a. Tests are still a work in progress.

While writing the test for `FindFreeDirections` I found out that the indexes for blocked cells are not correctly wrapping around the histogram.
The issues are present when the moving window azimuth index is greater than the grid and when the moving window elevation is smaller than 0.

Before:
error_polar
After:
coorect_polar

### CODE REVIEW

**Comment by nicovanduijn** [comment]
Love the visualization. What does the red color signify? Any reason there is no red 1 in the bottom right? (center of that blue cluster)

**Comment by mrivi** [comment]
red is the rejected cell, the one to which the obstacle was mapped to. 
blue is a blocked cell. It's with in the safety radius of a rejected cell. 
green is a candidate cell which is free from obstacles

**Comment by baumanta** [comment]
@mrivi yes, you are right. There is also a bug in the common function cartesian to polar and polar to cartesian. Those functions are not doing what they are supposed to. I plan on fixing that ASAP.

**Comment by mrivi** [comment]
@baumanta ok, I'll finish the tests to make this PR complete. We should fly tests these fixes, right?

**Comment by jkflying** [comment]
Looks good. Just a general comment, shouldn't the safety radius get wider as it gets higher/lower in the histogram, since it is in polar coordinates? Something like having a width of `radius/cos(elevation_angle)`?

**Comment by nicovanduijn** [comment]
I'm still not sure I understand how this histogram works. Particularly:
- It makes sense to wrap horizontally, since -180 is  the same azimuth direction as +180, but I don't understand why there is a wrapping in elevation. An obstacle above shouldn't wrap to an obstacle below the drone, right?
- The red 1 seems to be duplicated on the "left" obstacle but missing on the right one. Almost makes it look like the wrapping across the top happens diagonally
- I agree with julian that the radius should depend on elevation. Think about it like this: if the obstacle is straight above, it should cover the entire top row of the histogram (all azimuth angles)

**Comment by mrivi** [comment]
There is still a bug in the wrapping elevation. I have fixed it such that it wraps 180 in azimuth. Here's an example with a resolution of 20deg
screenshot from 2019-01-16 15-09-24

But there are still some corner cases. I am trying to fix those.

**Comment by nicovanduijn** [comment]
Oh, now I get it. You're totally right. I thought the wrapping was top-to-bottom, which of course wouldn't make sense. You just happened to show two obstacles in the same azimuth direction, which mislead me a bit.

**Comment by mrivi** [comment]
It should be working correctly now

File: local_planner/CMakeLists.txt [hunk 1]
```diff
@@ -199,13 +199,14 @@ if(CATKIN_ENABLE_TESTING)
 	catkin_add_gtest(${PROJECT_NAME}-test test/main.cpp
 	                                      test/test_example.cpp
 	                                      test/test_common.cpp
-	                                      test/test_local_planner.cpp)
+	                                      test/test_local_planner.cpp
+	                                      test/test_planner_functions.cpp)
```
Comment (jkflying) [inline local_planner/CMakeLists.txt]: This needs a rebase, there are new tests in `test_planner_functions.cpp` in master now.

File: local_planner/src/nodes/planner_functions.cpp [hunk 2]
```diff
@@ -399,19 +403,26 @@ void findFreeDirections(
 
           // Elevation index < 0
           if (i < 0 && j >= 0 && j < z_dim) {
-            a = -i;
-            b = z_dim - j - 1;
+            a = 0;
```
Comment (jkflying) [inline local_planner/src/nodes/planner_functions.cpp]: This logic gets a bit complicated, hopefully when we take into account the widths we can make a simpler version, maybe by switching to cartesian coordinates and back.

File: local_planner/test/test_planner_functions.cpp [hunk 3]
```diff
@@ -0,0 +1,560 @@
...
+  calculateFOV(h_fov, v_fov, z_FOV_idx_z_greater_grid_length, e_FOV_min,
+               e_FOV_max, yaw_z_greater_grid_length, pitch);
+  calculateFOV(h_fov, v_fov, z_FOV_idx_z_max_greater_grid, e_FOV_min, e_FOV_max,
+               yaw_z_max_greater_grid, pitch);
+  calculateFOV(h_fov, v_fov, z_FOV_idx3_z_min_smaller_zero, e_FOV_min,
+               e_FOV_max, yaw_z_min_smaller_zero, pitch);
+  calculateFOV(h_fov, v_fov, z_FOV_idx_z_smaller_zero, e_FOV_min, e_FOV_max,
+               yaw_z_smaller_zero, pitch);
+
+  // THEN:
```
Comment (jkflying) [inline local_planner/test/test_planner_functions.cpp]: Could you give a little explanation for what the checks do here? What are the different cases?

File: local_planner/test/test_planner_functions.cpp [hunk 4]
```diff
@@ -78,3 +79,556 @@ TEST(PlannerFunctions, generateNewHistogramSpecificCells) {
...
+  findFreeDirections(histogram, safety_radius, path_candidates, path_selected,
+                     path_rejected, path_blocked, path_waypoints,
+                     cost_path_candidates, goal, position, position_old,
+                     goal_cost_param, smooth_cost_param,
+                     height_change_cost_param_adapted, height_change_cost_param,
+                     only_yawed, resolution_alpha);
+
+  // THEN: we should have one rejected cell and the 8 neightbooring cells
+  // blocked
+  for (int i = 0, j = 0; i < path_blocked.cells.size(), j < blocked.size();
```
Comment (jkflying) [inline local_planner/test/test_planner_functions.cpp]: These test blocks seem to be used a lot, maybe they could be put in a function at the top of the file?

**Comment by mrivi** [comment]
I think this PR needs outdoor tests before being merged

**Comment by mrivi** [comment]
How should we go about merging this one? My testing capabilities are quite limited since I cannot run the simulation without timeouts on my laptop.

**Comment by jkflying** [comment]
Why don't you rebase this, and I'll test it in SITL for you tomorrow.

**Comment by mrivi** [comment]
@jkflying did you get a chance to test it?

**Comment by jkflying** [comment]
SITL performance looks the same as master, feel free to merge.

**Comment by mrivi** [comment]
@jkflying thank you!

**Pull request merged**